compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getPreviousSibling
+import org.jetbrains.kotlin.fir.declarations.FirTypeParameter
+import org.jetbrains.kotlin.fir.declarations.FirTypeParameterRefsOwner
+import org.jetbrains.kotlin.fir.expressions.FirGetClassCall
+import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.lexer.KtTokens.COLONCOLON
+import org.jetbrains.kotlin.lexer.KtTokens.QUEST
+
+object FirGetClassCallChecker : FirBasicExpressionChecker() {
+    override fun check(expression: FirStatement, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirGetClassCall) return
+        val source = expression.source ?: return
+        if (source.kind is FirFakeSourceElementKind) return
+
+        val argument = expression.argument as? FirResolvedQualifier ?: return
+        // Note that raw FIR drops marked nullability "?" in, e.g., `A?::class`, `A<T?>::class`, or `A<T?>?::class`.
+        // That is, AST structures for those expressions have token type QUEST, whereas FIR element doesn't have any information about it.
+        //
+        // A?::class -> CLASS_LITERAL_EXPRESSION(REFERENCE_EXPRESSION QUEST COLONCOLON "class")
+        // A<T?>::class -> CLASS_LITERAL_EXPRESSION(REFERENCE_EXPRESSION TYPE_ARGUMENT_LIST COLONCOLON "class")
+        // A<T?>?::class -> CLASS_LITERAL_EXPRESSION(REFERENCE_EXPRESSION TYPE_ARGUMENT_LIST QUEST COLONCOLON "class")
+        //   where TYPE_ARGUMENT_LIST may have QUEST in it
+        //
+        // Finding `child` of token type QUEST can't distinguish those, though. We could filter out QUEST from TYPE_ARGUMENT_LIST, but it
```

Comment:
Instead of finding the previous sibling of COLONCOLON, would it also work to call
`source.getChild(QUEST, depth = 1)`? It seems that the invalid cases would definitely have QUEST as a direct child but limiting the maximum depth would ensure you would not find QUEST in TYPE_ARGUMENT_LIST.

Reply:
You're totally right! It's much simpler now, thanks!